File: ruby/perfect-numbers/perfect_numbers.rb
module PerfectNumber
  def self.classify(n)
    raise RuntimeError, "n must be positive" unless n.positive?
    
    aliquot_sum = (1...n).filter { |d| n % d == 0 }.sum

    case aliquot_sum
    when n
      "perfect"
    when 0...n
      "deficient"
    else
      "abundant"
    end
  end
end

File: clojure/perfect-numbers/src/perfect_numbers.clj
(ns perfect-numbers)

(defn- factor-for? [n factor]
  (zero? (mod n factor)))

(defn- factors [n]
  (filter (partial factor-for? n) (range 1 (inc (quot n 2)))))

(defn- aliquot-sum [n]
  (reduce + (factors n)))

(defn classify [n]
  (when-not (pos? n)
    (throw (IllegalArgumentException. "n must be positive")))
  (let [aliquot (aliquot-sum n)]
    (cond
      (= aliquot n) :perfect
      (> aliquot n) :abundant
      :else         :deficient)))
